bowtie/_cli.py

```diff
@@ -1671,6 +1697,617 @@ async def info(
+        TextColumn("[bold blue]{task.description}"),
+        SpinnerColumn(finished_text=""),
+        BarColumn(bar_width=None),
+        TextColumn(
+            "[progress.percentage]{task.percentage:>3.0f}%",
+        ),
+        "•",
+        DownloadColumn(),
+        "•",
+        TimeElapsedColumn(),
+        console=console.Console(),
+        transient=True,
+    )
+    task = progress.add_task(
+        description=f"Fetching versions metadata of {id}",
+        total=None,
+    )
+
+    async with httpx.AsyncClient(timeout=10) as client:
+        try:
+            with progress:
+                url = (
+                    HOMEPAGE / "implementations" / id / "matrix-versions.json"
+                )
+                response = await client.get(str(url))
+                response.raise_for_status()
+
+                content = response.content
+                content_length = len(content)
+
+                progress.update(
+                    task,
+                    description=(
+                        f"Successfully fetched all versions of {id}!"
+                    ),
+                    completed=content_length,
+                    total=content_length,
+                    advance=content_length,
+                )
+                return frozenset(json.loads(content))
+        except httpx.HTTPStatusError as err:
+            progress.update(
+                task,
+                description=(
+                    f"[bold red]Could not fetch versions metadata of "
+                    f"{id}: {err.response.status_code}"
+                ),
+                completed=None,
+                total=None,
+                advance=None,
+            )
+            return frozenset()
+
+
+async def download_reports_for(
+    id: ConnectableId,
+    versions: Set[str],
+    dialects: Iterable[Dialect],
+) -> Iterable[tuple[str, Dialect, Response | None]]:
+    pretty_names_str = pretty_names_str_for(dialects)
+
+    progress = Progress(
+        TextColumn("[bold blue]{task.description}"),
+        SpinnerColumn(finished_text=""),
+        BarColumn(bar_width=None),
+        TextColumn("[progress.percentage]{task.percentage:>3.0f}%"),
+        "•",
+        MofNCompleteColumn(),
+        "•",
+        TimeElapsedColumn(),
+        console=console.Console(),
+        transient=True,
+    )
+
+    if versions:
+        total_files = len(versions) * len(list(dialects))
+        actual_downloaded_files = 0
+
+        task = progress.add_task(
+            description=(
+                "Preparing to download versioned reports "
+                f"of {id} for {pretty_names_str}"
+            ),
+            total=total_files,
+        )
+
+        async with httpx.AsyncClient(timeout=10) as client:
+
+            async def download_versioned_report_for(
+                version: str,
+                dialect: Dialect,
+            ):
+                try:
+                    url = (
+                        HOMEPAGE
```

Comment:
These kinds of details (what path the report lives at) should get encapsulated inside `Implementation` -- possibly by having a method that returns an async iterator which yields all the reports, though I think as I mentioned elsewhere maybe that makes progress reporting harder, but some of that I think is due to what I just mentioned -- you "can't" report progress in this way if it means you're uselessly waiting for all responses before getting along with other stuff (parsing).

Reply:
> You did say earlier to not use rich specific stuff in _core.py

Also this.

Reply:
I'm definitely not suggesting you do any rich things in `_core`, which is why I was focused simply on the logic and encapsulation bits, not on presentation.

There is way too much code here to go in the CLI as opposed to somewhere upstream of it -- I'll try to have a look to give more specific suggestions.

Reply:
Okay.